service/switchboardapp/src/test/java/com/switchboard/app/DecoderControllerTest.java

```diff
@@ -0,0 +1,114 @@
+import com.switchboard.app.domain.DecoderEntity;
+import com.switchboard.app.domain.DeviceEntity;
+import com.switchboard.app.exceptions.DeviceNotFoundException;
+import com.switchboard.app.fixture.DecoderFixture;
+import com.switchboard.app.fixture.DeviceFixture;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+import org.springframework.http.ResponseEntity;
+import org.springframework.mock.web.MockHttpServletRequest;
+import org.springframework.web.context.request.RequestContextHolder;
+import org.springframework.web.context.request.ServletRequestAttributes;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.Mockito.when;
+
+class DecoderControllerTest {
+
+    @InjectMocks
+    private DecoderController decoderController;
+
+    @Mock
+    private DecoderDaoImpl decoderService;
+
+    @Mock
+    private DeviceDaoImpl deviceService;
+
+    //stubbed Objects
+    static private DeviceEntity device1, device2;
+    static private DecoderEntity decoder1, decoder2;
+    static private List<DecoderEntity> listOfDecoders;
+
+    @BeforeAll
+    static void decoderFixture(){
+        //stubbing device and decoder objects
+        device1 = DeviceFixture.getDevice1();
+        device2 = DeviceFixture.getDevice2();
+        decoder1 = DecoderFixture.getDecoder1(device1);
+        decoder2 = DecoderFixture.getDecoder2(device2);
+        listOfDecoders = DecoderFixture.getListOfDecoders(decoder1, decoder2);
+    }
+
+    @BeforeEach
+    void setup(){
+        MockitoAnnotations.initMocks(this); //to be able to initiate decoderController object
+    }
+
+    @Test
+    final void testRetrieveAllDecoders(){
+        when(decoderService.getDecoders()).thenReturn(listOfDecoders);
+
+        List allDecoders = decoderController.retrieveAllDecoders();
+
+        assertFalse(allDecoders.isEmpty(),"allDecoders list is empty."); //check if an empty list was returned
+        assertIterableEquals(listOfDecoders, allDecoders,"listOfDecoders and allDecoders lists are not equal."); //check both lists contents
+    }
+
+    //When a decoder is available in the DB
+    @Test
+    final void testRetrieveDecoder(){
+        when(decoderService.findDecoder("1")).thenReturn(java.util.Optional.of(decoder1));
+
+        DecoderEntity actualDecoder = decoderController.retrieveDecoder("1").getContent();
+
+        assertNotNull(actualDecoder, "actualDecoder object is null.");
+        assertEquals(decoder1, actualDecoder, "expectedDecoder and actualDecoder objects are not equal.");
+    }
+
+    //When a decoder is unavailable in the DB
+    @Test
+    final void testRetrieveDecoderEmpty(){
+        assertThrows(DeviceNotFoundException.class, () -> {
```

Comment:
`DeviceNotFoundException.class` needs to change to `ExceptionType.DeviceNotFoundException.class`

If not changed, we will get a compile error on merge.

Reply:
Has been updated